src/query/api/v1/handler/prom/prom.go

```diff
@@ -38,29 +39,69 @@ func init() {
 	promql.SetDefaultEvaluationInterval(time.Minute)
 }
 
-// Options defines options for PromQL handler.
-type Options struct {
-	PromQLEngine *promql.Engine
+// opts defines options for PromQL handler.
+type opts struct {
+	promQLEngine *promql.Engine
+	instant      bool
+	newQueryFn   NewQueryFn
+}
+
+type Option interface {
+	apply(*opts) error
+}
+
+func WithEngine(promQLEngine *promql.Engine) Option {
+	return instantEngineOption{promQLEngine: promQLEngine, instant: false}
+}
+
+func WithInstantEngine(promQLEngine *promql.Engine) Option {
+	return instantEngineOption{promQLEngine: promQLEngine, instant: true}
+}
+
+type instantEngineOption struct {
+	promQLEngine *promql.Engine
+	instant      bool
+}
+
+func (o instantEngineOption) apply(options *opts) error {
+	if o.promQLEngine == nil {
+		return errors.New("invalid engine")
+	}
+	options.instant = o.instant
+	options.promQLEngine = o.promQLEngine
+	options.newQueryFn = newRangeQueryFn
+	if o.instant {
+		options.newQueryFn = newInstantQueryFn
+	}
+	return nil
+}
+
+func newDefaultOptions(hOpts options.HandlerOptions) opts {
+	return opts{
+		promQLEngine: hOpts.PrometheusEngine(),
+		instant:      false,
+		newQueryFn:   newRangeQueryFn,
+	}
```

Comment:
If we're going to do parameterized options for this change, we can simplify this a bit (also note the `WithFoo` exported symbols need comments):
```go
type opts struct {
  promQLEngine *promql.Engine
  instant      bool
  newQueryFn   NewQueryFn
}

// Option is a Prometheus handler option.
type Option func(*opts)

// WithEngine sets the PromQL engine.
func WithEngine(promQLEngine *promql.Engine) Option {
  return withEngine(promQLEngine, false)
}

// WithInstantEngine sets the PromQL instant engine.
func WithInstantEngine(promQLEngine *promql.Engine) Option {
  return withEngine(promQLEngine, true)
}

func withEngine(promQLEngine *promql.Engine, instant bool) Option {
  return func(o *opts) error {
    if promQLEngine == nil {
      return errors.New("invalid engine")
    }
    o.instant = o.instant
    o.promQLEngine = o.promQLEngine
    o.newQueryFn = newRangeQueryFn
    if instant {
      o.newQueryFn = newInstantQueryFn
    }
    return nil
  }
}

```

Reply:
Looks much simpler to me 👍🏻